### ISSUE TITLE
lastindex not defined for CartesianIndex

### ISSUE THREAD

**Issue by colintbowers** [JuliaLang/julia#31815]
I understand `CartesianIndex` is deliberately not iterable. However, it also doesn't have a `lastindex` method. Is this deliberate? It feels like something that would be useful.

For example (on v1.1):
```
julia> x = CartesianIndex{2}(3, 4)
CartesianIndex(3, 4)

julia> lastindex(x)
ERROR: MethodError: no method matching lastindex(::CartesianIndex{2})
Closest candidates are:
  lastindex(::Core.SimpleVector) at essentials.jl:567
  lastindex(::Cmd) at process.jl:940
  lastindex(::Base64.Buffer) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.1/Base64/src/buffer.jl:19
  ...
Stacktrace:
 [1] top-level scope at none:0
```

**Comment by timholy** [comment]
Given that it's indexable that seems like an oversight. PR?

BTW, you don't need the `{2}` in your example.

**Comment by martinholters** [comment]
We do have `length`, though. OTOH we don't have `axes` or `eachindex`. Should those be added, too?

**Comment by timholy** [comment]
`eachindex` seems reasonable. I worry that `axes` would move us too far towards iterability.

**Comment by mbauman** [comment]
I agree that we should have `firstindex` and `lastindex`. I'm not as sure about `eachindex` and `axes` — not only do they bring us closer to iterability but they also bring us closer to broadcastability.  I wouldn't implement `eachindex` for the sake of implementing it; instead I'd wait for an organic feature request that comes from a real use case.

**Comment by stevengj** [comment]
I think if a type supports `getindex`, then as a matter of general principle it should support all of the `*index` functions.

**Comment by mbauman** [comment]
I agree in general, but this case definitely isn't a general case.  It's an object we want to behave like a scalar (and not a container) but have chosen to use indexing as the API to access its fields.  That said, I do see that strings implement `keys` and `eachindex` (but not `axes`), so perhaps that is indeed where this line should fall.

**Comment by adienes** [comment]
added in https://github.com/JuliaLang/julia/pull/57985
